algo.lua

```diff
@@ -0,0 +1,137 @@
+
+        local algType = ffi.new(algo_t, 1)
+
+        if cudnn.benchmark or cudnn.fastest then -- the manual auto-tuner is run
+            if autotunerCache[1][self.autotunerHash] then
+                algType[0] = autotunerCache[1][self.autotunerHash]
+                if cudnn.verbose then
+                   print('\n', findAPI, ' using cached algo = ' , algType[0] , ' for: ', self.autotunerHash)
+                end
+            else
+                local perfResults = ffi.new(perf_t, 1)
+                local intt = torch.IntTensor(1)
+                errcheck(findAPI,
+                         cudnn.getHandle(),
+                         params[1], params[2], params[3], params[4],
+                         1, intt:data(), perfResults)
+                algType[0] = perfResults[0].algo
+                autotunerCache[1][self.autotunerHash] = perfResults[0].algo
+                if cudnn.verbose then
+                    print(string.format(
+                              "\n" .. findAPI .. " Time: %3.5f Memory: %8d Algorithm: %d"
+                                  .. " hash: %45s",
+                              perfResults[0].time, tonumber(perfResults[0].memory),
+                              tonumber(perfResults[0].algo), self.autotunerHash ))
+
+                end
+            end
+        else
+
+           local algWorkspaceLimit = self.workspace_limit
+              or (self.nInputPlane * self.kH * self.kW * self.weight.elementSize())
+
+            errcheck(getAPI,
+                     cudnn.getHandle(),
+                     params[1], params[2], params[3], params[4],
+                     algSearchMode, algWorkspaceLimit, algType)
+                if cudnn.verbose then
+                   print(string.format(
+                     "\n" .. getAPI .. " Limit: %d Algorithm: %d",
+                     tonumber(algWorkspaceLimit),
+                     tonumber(algType[0])))
+                end
+        end
+        local bufSize = torch.LongTensor(1)
+        errcheck(wsAPI,
+                 cudnn.getHandle(),
+                 params[1], params[2], params[3], params[4],
+                 algType[0], bufSize:data())
+
+        self.extraBuffer = self.extraBuffer or cudnn.getSharedWorkspace()
+        local extraBufferSizeInBytes = self.extraBuffer:nElement() * self.extraBuffer.elementSize()
+
+       if cudnn.verbose then
+           print(string.format(
+                    "\n" .. wsAPI .. " returned bufSize: %d, current extraBufferSizeInBytes: %d, %d elements",
+                    tonumber(bufSize[1]), tonumber(extraBufferSizeInBytes), tonumber(self.extraBuffer:nElement())))
+        end
+
+        if extraBufferSizeInBytes < bufSize[1] then
+           self.extraBuffer:resize(math.ceil(bufSize[1]/self.extraBuffer.elementSize()))
+        end
+        return algType[0]
+end
+
+function algo.prepareHash(self, input_slice, output_slice)
+   local function shape(x)
```

Comment:
Thanks @ngimel ! I will use that one liner from VolumetricConvolution and then refactor it, too :). Any ideas why I may not have seen DoubleTensor errors fixed as you did ?